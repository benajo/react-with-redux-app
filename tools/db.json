{
  "courses": [
    {
      "id": 1,
      "title": "Securing React Apps with Auth0.1",
      "slug": "react-auth0-authentication-security",
      "authorId": 1,
      "category": "JavaScript"
    },
    {
      "id": 2,
      "title": "React: The Big Picture",
      "slug": "react-big-picture",
      "authorId": 1,
      "category": "JavaScript"
    },
    {
      "id": 3,
      "title": "Creating Reusable React Components",
      "slug": "react-creating-reusable-components",
      "authorId": 1,
      "category": "JavaScript"
    },
    {
      "id": 4,
      "title": "Building a JavaScript Development Environment",
      "slug": "javascript-development-environment",
      "authorId": 1,
      "category": "JavaScript"
    },
    {
      "id": 5,
      "title": "Building Applications with React and Redux",
      "slug": "react-redux-react-router-es6",
      "authorId": 1,
      "category": "JavaScript"
    },
    {
      "id": 6,
      "title": "Building Applications in React and Flux",
      "slug": "react-flux-building-applications",
      "authorId": 1,
      "category": "JavaScript"
    },
    {
      "id": 7,
      "title": "Clean Code: Writing Code for Humans",
      "slug": "writing-clean-code-humans",
      "authorId": 1,
      "category": "Software Practices"
    },
    {
      "id": 8,
      "title": "Architecting Applications for the Real World",
      "slug": "architecting-applications-dotnet",
      "authorId": 1,
      "category": "Software Architecture"
    },
    {
      "id": 9,
      "title": "Becoming an Outlier: Reprogramming the Developer Mind",
      "slug": "career-reboot-for-developer-mind",
      "authorId": 1,
      "category": "Career"
    },
    {
      "id": 10,
      "title": "Web Component Fundamentals",
      "slug": "web-components-shadow-dom",
      "authorId": 1,
      "category": "HTML5"
    },
    {
      "id": 11,
      "title": "test",
      "authorId": 1,
      "category": "testcat",
      "createdAt": 1569005636631,
      "slug": "test"
    },
    {
      "id": 12,
      "title": "test2",
      "authorId": 3,
      "category": "asd",
      "createdAt": 1569005961845,
      "slug": "test2"
    },
    {
      "id": 13,
      "title": "jjj",
      "authorId": 3,
      "category": "jjjjkk",
      "createdAt": 1569006187120,
      "slug": "jjj"
    },
    {
      "id": 14,
      "title": "hello",
      "authorId": 1,
      "category": "you",
      "createdAt": 1569006296726,
      "slug": "hello"
    },
    {
      "id": 15,
      "title": "foo",
      "authorId": 1,
      "category": "bar",
      "createdAt": 1569006310621,
      "slug": "foo"
    },
    {
      "id": 16,
      "title": "foo",
      "authorId": 1,
      "category": "bar",
      "createdAt": 1569006318445,
      "slug": "foo"
    }
  ],
  "authors": [
    {
      "id": 1,
      "name": "Cory House"
    },
    {
      "id": 2,
      "name": "Scott Allen"
    },
    {
      "id": 3,
      "name": "Dan Wahlin"
    }
  ]
}