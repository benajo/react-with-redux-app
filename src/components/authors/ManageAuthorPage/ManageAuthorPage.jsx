import React, { useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import PropTypes from 'prop-types';
import { toast } from 'react-toastify';
import { loadAuthors, saveAuthor } from '../../../redux/actions/authorActions';
import AuthorForm from '../AuthorForm';
import { newAuthor } from '../../../../tools/mockData';
import Spinner from '../../common/Spinner';

const getAuthorBySlug = (authors, slug) => authors.find((author) => author.slug === slug) || null;

const ManageAuthorPage = ({ history, match }) => {
  const dispatch = useDispatch();

  const { authors, authorsLoaded } = useSelector((state) => state);

  const [author, setAuthor] = useState();
  const [errors, setErrors] = useState();
  const [saving, setSaving] = useState(false);

  useEffect(() => {
    dispatch(loadAuthors()).catch((error) => {
      toast.error(`Loading authors failed: ${error}`);
    });
  }, [authorsLoaded]);

  useEffect(() => {
    const { slug } = match.params;
    const getAuthor = slug && authors && authors.length > 0
      ? getAuthorBySlug(authors, slug)
      : newAuthor;
    setAuthor({ ...getAuthor });
  }, [authors]);

  const handleChange = (event) => {
    const { name, value } = event.target;
    setAuthor((prevAuthor) => ({
      ...prevAuthor,
      [name]: name === 'authorId' ? parseInt(value, 10) : value,
    }));
  };

  const formIsValid = () => {
    const { name } = author;
    const errs = {};

    if (!name) errs.name = 'Name is required.';

    setErrors(errs);
    // Form is valid if the errors object still has no properties
    return Object.keys(errs).length === 0;
  };

  const handleSave = (event) => {
    event.preventDefault();
    if (!formIsValid()) return;
    setSaving(true);
    dispatch(saveAuthor(author))
      .then(() => {
        toast.success('Author saved.');
        history.push('/authors');
      })
      .catch((error) => {
        setSaving(false);
        setErrors({ onSave: error.message });
      });
  };

  return !author || !authors ? (
    <Spinner />
  ) : (
    <>
      <AuthorForm
        author={author}
        errors={errors}
        authors={authors}
        onChange={handleChange}
        onSave={handleSave}
        saving={saving}
      />
    </>
  );
};

ManageAuthorPage.propTypes = {
  history: PropTypes.shape({
    push: PropTypes.func.isRequired,
  }).isRequired,
  match: PropTypes.shape({
    params: PropTypes.shape({
      slug: PropTypes.string,
    }),
  }).isRequired,
};

export default ManageAuthorPage;
