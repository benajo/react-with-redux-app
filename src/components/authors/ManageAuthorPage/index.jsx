import ManageAuthorPage from './ManageAuthorPage';

export default ManageAuthorPage;
