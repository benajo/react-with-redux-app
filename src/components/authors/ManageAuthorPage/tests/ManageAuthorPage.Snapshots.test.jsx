import React from 'react';
import renderer from 'react-test-renderer';
import * as ReactRedux from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import ManageAuthorPage from '../ManageAuthorPage';
import { authors } from '../../../../../tools/mockData';

const middleware = [thunk];
const mockStore = configureMockStore(middleware);

let spyOnUseDispatch;

describe('ManageAuthorPage', () => {
  beforeEach(() => {
    const dispatch = jest.fn(() => Promise.resolve());
    spyOnUseDispatch = jest
      .spyOn(ReactRedux, 'useDispatch')
      .mockReturnValue(dispatch);
  });

  afterEach(() => {
    spyOnUseDispatch.mockRestore();
  });

  it('render ManageAuthorPage with empty authors', () => {
    // const store = mockStore({
    //   authors: [],
    // });
    // const tree = renderer.create(
    //   <ReactRedux.Provider store={store}>
    //     <ManageAuthorPage />
    //   </ReactRedux.Provider>,
    // );
    expect('').toMatchSnapshot();
  });
});
