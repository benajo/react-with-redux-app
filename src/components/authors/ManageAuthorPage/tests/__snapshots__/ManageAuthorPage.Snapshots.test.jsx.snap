// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ManageAuthorPage render ManageAuthorPage with empty authors 1`] = `""`;
